presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java

```diff
@@ -0,0 +1,215 @@
+                .toArray(new TablestoreColumnHandle[columnHandles.size()]);
+        this.split = split;
+        this.tableHandle = split.getTableHandle();
+        this.tableName = requireNonNull(tableHandle.getTableName(), "tableName is null");
+        this.primaryKeys = columnHandles.stream()
+                .filter(TablestoreColumnHandle::isPrimaryKey)
+                .sorted(Comparator.comparingInt(TablestoreColumnHandle::getPkPosition))
+                .collect(Collectors.toList());
+        this.enableLooseCast = TablestoreFacade.enableLooseCast(session);
+        this.startTimestamp = currentTimeMillis();
+    }
+
+    protected <T> T callAndRecord(int field, Callable<T> call)
+    {
+        long t1 = nanoTime();
+        try {
+            checkStates(field);
+            return call.call();
+        }
+        catch (Throwable t) {
+            throw new RuntimeException("row[pk=" + currentPrimaryKeyAsDebug() + "] field[index=" + field + "] -> " + t.getMessage(), t);
+        }
+        finally {
+            readNano += nanoTime() - t1;
+        }
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        checkClosed();
+        return (currentTimeMillis() - startTimestamp) * 1000_000;
+    }
+
+    protected long getTotalCostMillis()
+    {
+        return getReadTimeNanos() / 1000_000;
+    }
+
+    protected long getApiCostMillis()
+    {
+        return apiNano / 1000_000;
+    }
+
+    protected long getReadCostMillis()
+    {
+        return readNano / 1000_000;
+    }
+
+    @Override
+    public Type getType(int field)
+    {
+        checkStates(field);
+        return columnHandles[field].getColumnType();
+    }
+
+    protected String getColumnName(int idx)
+    {
+        return columnHandles[idx].getColumnName();
+    }
+
+    protected boolean isPrimaryKey(int idx)
+    {
+        return columnHandles[idx].isPrimaryKey();
+    }
+
+    protected abstract boolean hasNext();
+
+    protected abstract void next();
+
+    /**
+     * 为了出错时好定位问题行
+     */
+    protected abstract String currentPrimaryKeyAsDebug();
+
+    protected String toString4Value(Object value)
+    {
+        if (value == null) {
+            return "null";
+        }
+        else if (value instanceof byte[]) {
+            String x = Base64.getEncoder().encodeToString((byte[]) value);
+            return x.length() > 24 ? x.substring(0, 24) + "..." : x;
+        }
+        else {
+            return String.valueOf(value);
+        }
+    }
+
+    @Override
+    public boolean advanceNextPosition()
+    {
+        checkClosed();
+        String queryId = session.getQueryId();
+        String fmt = "queryId=%s scannedRows=%s elapsedMills=%d msg=%s";
+
+        try {
+            long t1 = nanoTime();
+            if (hasNext()) {
+                next(); //有可能报错
+                long delta = System.nanoTime() - t1;
+
+                if (delta >= 10 * 1_000_000) {
```

Comment:
Why we have 10 times 1_000_000? should we have a constant here?

Reply:
1. added a constant.
2. It is to make sure we only log every 10ms -- so it will not generate too much logs.